### PR TITLE
ALTAPPS-1092: Shared interview preparation widget

### PR DESCRIPTION
**Pull request by XanderZhu**
**YouTrack Issues**:
[ALTAPPS-1092](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-1092)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] Sentry Performance Monitoring of screen loading is added for new screens;
- [x] View analytics events are added for new screens;
- [x] New analytics events are documented;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

**Description**
- Implement feature just for interview widget on the main screen

### CODE REVIEW

**Comment by XanderZhu** [comment]
@ivan-magda 
iOS build failed with a strange error
```
Looks like fastlane ran into a build/archive error with your project
[13:27:08]: It's hard to tell what's causing the error, so we wrote some guides on how
[13:27:08]: to troubleshoot build and signing issues: https://docs.fastlane.tools/codesigning/getting-started/
```

**Comment by XanderZhu** [comment]
@ivan-magda 
Review this PR first

File: shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionReducer.kt [hunk 1]
```diff
@@ -227,8 +201,73 @@ class StepCompletionReducer(private val stepRoute: StepRoute) : StateReducer<Sta
...
+    private fun handleStepSolved(
+        state: State,
+        message: Message.StepSolved
+    ): StepCompletionReducerResult =
+        if (message.stepId == state.currentStep.id) {
+            when (stepRoute) {
+                is StepRoute.Learn ->
+                    state to setOf(Action.UpdateProblemsLimit)
+                is StepRoute.InterviewPreparation ->
+                    state to setOf(InternalAction.MarkInterviewStepAsSolved(message.stepId))
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionReducer.kt:264]: we also should update limits after interview step is solved

File: shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionActionDispatcher.kt [hunk 2]
```diff
@@ -267,4 +279,34 @@ class StepCompletionActionDispatcher(
...
+                        SharedResources.strings.step_theory_failed_to_continue_practicing
+                    )
+                )
+            }
+        ) {
+            val nextInterviewStepId = interviewStepsStateRepository
+                .getState(forceUpdate = false)
+                .getOrThrow()
+                .shuffled()
+                .firstOrNull()
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionActionDispatcher.kt:300]: Check shared task description:

1. We should shuffle state repository on widget click
2. Get last step

File: shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionActionDispatcher.kt [hunk 3]
```diff
@@ -267,4 +279,34 @@ class StepCompletionActionDispatcher(
...
+                nextInterviewStepId?.id?.let(StepRoute::InterviewPreparation)
+            )
+        }.let(onNewMessage)
+    }
+
+    private suspend fun handleMarkInterviewStepAsSolved(action: InternalAction.MarkInterviewStepAsSolved) {
+        interviewStepsStateRepository.updateState { steps ->
+            steps.mutate { remove(StepId(action.stepId)) }
+        }
+    }
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionActionDispatcher.kt:311]: I saw the same code in the `InterviewPreparationWidgetActionDispatcher`, you are trying to mutate state twice.

I suggest you to subscribe for `InterviewStepsStateRepository` changes in the `InterviewPreparationWidgetActionDispatcher` and update `InterviewPreparationWidgetFeature.State.Content.steps` directly with new state

File: shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionReducer.kt [hunk 4]
```diff
@@ -227,8 +201,73 @@ class StepCompletionReducer(private val stepRoute: StepRoute) : StateReducer<Sta
...
+    private fun handleFetchNextInterviewStepResultSuccess(
+        state: State,
+        message: InternalMessage.FetchNextInterviewStepResultSuccess
+    ): StepCompletionReducerResult =
+        state.copy(isPracticingLoading = false) to
+            setOf(
+                if (message.newStepRoute != null) {
+                    Action.ViewAction.ReloadStep(message.newStepRoute)
+                } else {
+                    Action.ViewAction.NavigateTo.Back
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionReducer.kt:251]: When `message.newStepRoute == null` than I suppose we should show a modal with success message:

> _При решении задачи и обработки этого события в StepCompletionFeature добавить проверку - если StepRoute это InterviewPreparation, то удалять последнюю задачу из InterviewStepsStateRepository и открывать последнюю после удаления. Если после удаления последней задачи в списке не осталось задач - выбрасывать ViewAction, который покажет модалку с поздравлением решения задач_

Screenshot 2023-12-29 at 7 13 11 PM

File: shared/src/commonMain/kotlin/org/hyperskill/app/interview_preparation/domain/analytics/InterviewPreparationWidgetViewedHyperskillAnalyticsEvent.kt [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * JSON payload:
+ * ```
+ * {
+ *     "route": "/home/interview_preparation_widget",
+ *     "action": "view"
+ * }
+ * ```
+ * @see HyperskillAnalyticEvent
+ */
+object InterviewPreparationWidgetViewedHyperskillAnalyticsEvent : HyperskillAnalyticEvent(
```
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/interview_preparation/domain/analytics/InterviewPreparationWidgetViewedHyperskillAnalyticsEvent.kt:19]: @ivan-magda 
In the task description, there is a point about view-event for this widget. See the acceptance criteria

[the remark this answers is not in the document]
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/interview_preparation/domain/analytics/InterviewPreparationWidgetViewedHyperskillAnalyticsEvent.kt:19]: @XanderZhu thank you for explaining, I missed that 👍

File: shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/StepId.kt [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+import kotlin.jvm.JvmInline
+
+/**
+ * Represents a unique identifier for a step.
+ *
+ * @property id The ID of the step.
+ */
+@JvmInline
+value class StepId(val id: Long)
```
[the remark this answers is not in the document]
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/StepId.kt:11]: @XanderZhu I want to point out that this is very nice that you started such kind of improvement in our code base 👍

File: shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/StepRoute.kt [hunk 7]
```diff
@@ -30,13 +29,11 @@ sealed interface StepRoute {
 
     @Serializable
     data class LearnDaily(override val stepId: Long) : StepRoute {
-        @Transient
-        override val analyticRoute: HyperskillAnalyticRoute =
-            HyperskillAnalyticRoute.Learn.Daily(stepId)
+        override val analyticRoute: HyperskillAnalyticRoute
+            get() = HyperskillAnalyticRoute.Learn.Daily(stepId)
```
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/StepRoute.kt:33]: I decided to use get access to not use `Transient` annotation and make the serialization more clear. Some other step routes use the same way.
I don't think it affects performance in any way.

File: shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/StepRoute.kt [hunk 8]
```diff
@@ -63,12 +60,21 @@ sealed interface StepRoute {
         val projectId: Long,
         val stageId: Long,
     ) : StepRoute {
-        @Transient
```
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/StepRoute.kt:66]: Properties without backing-field are not serialized
https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/basic-serialization.md#backing-fields-are-serialized

File: shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionActionDispatcher.kt (hunk 2 above)
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionActionDispatcher.kt:300]: Honestly, I don't see the difference where to shuffle steps and what the step to take (first, or last).
Anyway, I'll move the shuffling out from the feature, to not modify state in the feature and in the repository (as you mentioned before).

File: shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/StepId.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/StepId.kt:11]: @ivan-magda 
`value class` is used specially for something like StepId to have a special type without a memory footprint.
So in this case replace a raw Long with a special StepId type.
I remove StepId, let's discuss it later in the debt task.

File: shared/src/commonMain/kotlin/org/hyperskill/app/sentry/domain/SentryExtentions.kt [hunk 9]
```diff
@@ -22,6 +23,8 @@ suspend inline fun <T> SentryInteractor.withTransaction(
         val result = measureBlock()
         finishTransaction(transaction)
         result
+    } catch (e: CancellationException) {
+        throw e
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/sentry/domain/SentryExtentions.kt:27]: @XanderZhu 

1. If `CancellationException` catched and then rethrows, could this lead to a runtime crash?
2. I suppose the Sentry transaction will never be finished

**Review by ivan-magda — APPROVED** [review]

File: shared/src/commonMain/kotlin/org/hyperskill/app/sentry/domain/SentryExtentions.kt (hunk 9 above)
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/sentry/domain/SentryExtentions.kt:27]: @ivan-magda 
> If CancellationException is caught and then rethrows, could this lead to a runtime crash?

No, it couldn't. `CancellationException` is a special exception that is used in coroutines to cancel running coroutines and it shouldn't be watched.
We use `CoroutineExceptionHandler` to handle all the exceptions inside the ActionDisptacher coroutines. (see [code](https://github.com/hyperskill/mobile-app/blob/6c05f421f1b6e768954f82ce77d0dcdf926b7613/shared/src/commonMain/kotlin/org/hyperskill/app/core/presentation/ActionDispatcherOptions.kt#L18-L22)). `CoroutineExceptionHandler` handles all the exceptions except of `CancellationException`.
https://kotlinlang.org/docs/exception-handling.html#supervision-job
https://stackoverflow.com/a/76261124

Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/sentry/domain/SentryExtentions.kt:27]: @ivan-magda 
> I suppose the Sentry transaction will never be finished

Good point. We should finish sentry transition for `CancellationException` too.

**Pull request merged**